olp-cpp-sdk-core/src/network2/winhttp/NetworkProtocolWinHttp.cpp

```diff
@@ -0,0 +1,1063 @@
+      requestsToCancel.push_back(request.second);
+    }
+    while (!requestsToCancel.empty()) {
+      std::shared_ptr<RequestData> request = requestsToCancel.front();
+      WinHttpCloseHandle(request->http_request);
+      request->http_request = NULL;
+      pendingResults.push_back(request->result_data);
+      requestsToCancel.erase(requestsToCancel.begin());
+    }
+  }
+
+  if (http_session_) {
+    WinHttpCloseHandle(http_session_);
+    http_session_ = NULL;
+  }
+
+  if (event_ != INVALID_HANDLE_VALUE) SetEvent(event_);
+  if (thread_ != INVALID_HANDLE_VALUE) {
+    if (GetCurrentThreadId() != GetThreadId(thread_)) {
+      WaitForSingleObject(thread_, INFINITE);
+    }
+  }
+  CloseHandle(event_);
+  CloseHandle(thread_);
+  thread_ = event_ = INVALID_HANDLE_VALUE;
+
+  {
+    std::unique_lock<std::recursive_mutex> lock(mutex_);
+    http_connections_.clear();
+    while (!results_.empty()) {
+      pendingResults.push_back(results_.front());
+      results_.pop();
+    }
+  }
+
+  for (auto& result : pendingResults) {
+    std::unique_lock<std::recursive_mutex> lock(mutex_);
+    if (result->user_callback) {
+      result->user_callback(
+          NetworkResponse()
+              .WithRequestId(result->request_id)
+              .WithStatus(static_cast<int>(ErrorCode::OFFLINE_ERROR))
+              .WithError("Offline"));
+      result->user_callback = nullptr;
+    }
+  }
+}
+
+NetworkStatus NetworkProtocolWinHttp::Send(
+    NetworkRequest request, std::shared_ptr<std::ostream> payload,
+    Callback callback, HeaderCallback header_callback,
+    DataCallback data_callback) {
+  RequestId id = recent_request_id_.fetch_add(1);
+
+  // if (!config->GetNetworkInterface().empty()) {
+  //  return olp::network::NetworkProtocol::
+  //      ErrorNetworkInterfaceOptionNotImplemented;
+  //}
+
+  // if (!config->GetCaCert().empty()) {
+  //  return olp::network::NetworkProtocol::ErrorCaCertOptionNotImplemented;
+  //}
+
+  URL_COMPONENTS urlComponents;
+  ZeroMemory(&urlComponents, sizeof(urlComponents));
+  urlComponents.dwStructSize = sizeof(urlComponents);
+  urlComponents.dwSchemeLength = (DWORD)-1;
+  urlComponents.dwHostNameLength = (DWORD)-1;
+  urlComponents.dwUrlPathLength = (DWORD)-1;
+  urlComponents.dwExtraInfoLength = (DWORD)-1;
+
+  std::wstring url(request.GetUrl().begin(), request.GetUrl().end());
+  if (!WinHttpCrackUrl(url.c_str(), (DWORD)url.length(), 0, &urlComponents)) {
+    EDGE_SDK_LOG_ERROR(LOGTAG, "WinHttpCrackUrl failed " << GetLastError());
+    // return olp::network::NetworkProtocol::ErrorInvalidRequest;
+    return NetworkStatus(ErrorCode::INVALID_URL_ERROR);
+  }
+
+  std::shared_ptr<RequestData> handle;
+  {
+    std::unique_lock<std::recursive_mutex> lock(mutex_);
+    std::wstring server(url.data(), size_t(urlComponents.lpszUrlPath -
+                                           urlComponents.lpszScheme));
+    std::shared_ptr<ConnectionData> connection = http_connections_[server];
```

Comment:
why not work directly with reference:
```suggestion
    auto& connection = http_connections_[server];
```

then you dont need to set it again at the end